lib/saluki-components/src/decoders/otlp/mod.rs

```diff
@@ -0,0 +1,173 @@
+        builder
+            .minimum()
+            .with_single_value::<OtlpTracesDecoder>("decoder struct");
+    }
+}
+
+/// OTLP traces decoder.
+pub struct OtlpTracesDecoder {
+    translator: OtlpTracesTranslator,
+    metrics: Metrics,
+}
+
+#[async_trait]
+impl Decoder for OtlpTracesDecoder {
+    async fn run(self: Box<Self>, mut context: DecoderContext) -> Result<(), GenericError> {
+        let mut health = context.take_health_handle();
+        health.mark_ready();
+
+        debug!("OTLP traces decoder started.");
+
+        // Set a buffer flush interval of 100ms to ensure we flush buffered events periodically.
+        let mut buffer_flush = interval(Duration::from_millis(100));
+        buffer_flush.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        let mut event_buffer_manager = EventBufferManager::default();
+
+        loop {
+            select! {
+                maybe_payload = context.payloads().next() => {
+                    let payload = match maybe_payload {
+                        Some(payload) => payload,
+                        None => {
+                            debug!("Payloads stream closed, shutting down decoder.");
+                            break;
+                        }
+                    };
+
+                    let grpc_payload = match payload.try_into_grpc_payload() {
+                        Some(grpc) => grpc,
+                        None => {
+                            error!("Received non-gRPC payload in OTLP decoder.");
+                            continue;
+                        }
+                    };
+
+                    match grpc_payload.service_path() {
+                path if path == &*OTLP_TRACES_GRPC_SERVICE_PATH => {
+                    let body = grpc_payload.body().clone();
+
+                    let request = match ExportTraceServiceRequest::decode(body) {
+                        Ok(req) => req,
+                        Err(e) => {
+                            error!(error = %e, "Failed to decode OTLP trace request.");
+                            continue;
+                        }
+                    };
+
+                    for resource_spans in request.resource_spans {
+                        let trace_events = self.translator.translate_resource_spans(resource_spans, &self.metrics);
+                        for trace_event in trace_events {
+                            if let Some(event_buffer) = event_buffer_manager.try_push(trace_event) {
+                                if let Err(e) = context.dispatcher().dispatch(event_buffer).await {
+                                    error!(error = %e, "Failed to dispatch trace events.");
+                                }
+                            }
+                        }
+                    }
+                }
+                path if path == &*OTLP_METRICS_GRPC_SERVICE_PATH => {
+                    error!("OTLP metrics decoding not yet implemented.");
```

Comment:
Same thing here about saying that we're dropping the payload:

`OTLP metrics decoding not yet implemented. Dropping metrics payload.`